cmake/githash.pl

```diff
@@ -0,0 +1,203 @@
+#!/usr/bin/perl -w
+
+#
+# Copyright (c) 2016 Facebook, Inc.
+# All rights reserved
+#
+
+#
+# Git hash/date header file generator
+#
+
+use strict;
+
+use Cwd;
+use File::Copy qw(copy);
+use Getopt::Long qw(GetOptions);
+use Pod::Usage qw(pod2usage);
+
+sub check_all_or_none {
+  my $bitmask = 0;
+
+  $bitmask = $bitmask | 1 if shift ne "";
+  $bitmask = $bitmask | 2 if shift ne "";
+  $bitmask = $bitmask | 4 if shift ne "";
+  $bitmask = $bitmask | 8 if shift ne "";
+
+  pod2usage("You must specify none or all of the git hashes and dates")
+      unless ($bitmask == 0 || $bitmask == 15);
+}
+
+# Function to retrieve the git hash and date from a repository
+sub git_hash_and_date {
+  my $subdir = shift;
+  my $orgdir = Cwd::getcwd;
+
+  # Switch directories to the specified one
+  chdir($subdir) or die "Can't change directory to '$subdir'";
+
+  # Get the hash and date from the most recent revision in the repository
+  my $git_cmd = "git log -1 --format=\"%H;%cI\"";
+  open (my $log, "$git_cmd |") or die "Can't run $git_cmd";
+
+  my $githash = "";
+  my $gitdate = "";
+
+  # Loop through all the lines - we should only have one line
+  while (<$log>) {
+    # Search for a line that has a hash, a semicolon and the date
+    if (/^([0-9a-f]{7,40});(.*)\n/) {
+      die "Unexpected multiple matching log lines" unless !length($githash);
+      $githash = $1;
+      $gitdate = $2;
+    }
+  }
+
+  # Make sure we got something
+  die "No matching log lines" unless length($githash);
+
+  # Close the input and switch back to the original subdirectory
+  close($log);
+  chdir($orgdir);
+
+  # Return the found hash and date
+  return ($githash, $gitdate);
+}
+
+# main function
+sub main {
+  my $root = "";
+  my $infile = "";
+  my $mysql_git_hash = "";
+  my $mysql_git_date = "";
+  my $rocksdb_git_hash = "";
+  my $rocksdb_git_date = "";
+  my $help = 0;
+
+  # Get the parameters
+  GetOptions(
+    'help|?' => \$help,
+    'git_root=s' => \$root,
+    'file=s' => \$infile,
+    'mysql_githash=s' => \$mysql_git_hash,
+    'mysql_gitdate=s' => \$mysql_git_date,
+    'rocksdb_githash=s' => \$rocksdb_git_hash,
+    'rocksdb_gitdate=s' => \$rocksdb_git_date
+  ) or pod2usage();
+
+  # Validate the parameters
+  pod2usage(-verbose => 1) if $help;
+  pod2usage("missing required parameter --git_root") if $root eq "";
+  pos2usage("missing required parameter --file") if $infile eq "";
+  check_all_or_none($mysql_git_hash, $mysql_git_date,
+                    $rocksdb_git_hash, $rocksdb_git_date);
+
+  my $rocksdb = "$root/rocksdb";
+  my $outfile = "$infile.tmp";
+
+  if ($mysql_git_hash eq "") {
+    # retrieve the git hash and date for the main repository
+    ($mysql_git_hash, $mysql_git_date) = git_hash_and_date $root;
+    # retrieve the git hash and date for the rocksdb submodule
```

Comment:
Should also have this check to see if the rocksdb hash have been set. If not, then perform the git_hash_and_date, which requires "cd rocksdb".